lib/util/Floats.v3

```diff
@@ -0,0 +1,665 @@
+// Copyright 2024 Virgil authors. All rights reserved.
+// See LICENSE for details of Apache 2.0 license.
+
+// ====================== Start of Ryu license ========================
+// Copyright 2018 Ulf Adams
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ======================== End of Ryu license =========================
+
+// ===========================  Credits  ===============================
+// This file is an combination of
+// 1. Implementations in Virgil with respect to the ryu paper
+//    @ https://dl.acm.org/doi/pdf/10.1145/3192366.3192369
+// 2. Heavily modified arithmetics and numerical methods derived
+//	  from of the original work by Ulf Adams
+//    @ https://github.com/ulfjack/ryu.
+// =====================================================================
+
+component Floats {
+	private def FLOAT_EXP_WIDTH: byte = 8;
+	private def FLOAT_SIGN_WIDTH: byte = 1;
+	private def FLOAT_MANTISSA_BITS: byte = 23;
+	private def FLOAT_BIAS: byte = 127;
+
+	private def DEBUG: bool = true;
+
+	def renderFloat(val: float, a: Array<byte>, pos: int) -> int {
+		var length: int = 0;
+		if (u32.view(val) >> byte.!(FLOAT_MANTISSA_BITS + FLOAT_EXP_WIDTH) == 1u) {
+			a[pos++] = '-';
+			return 1 + renderPosFloat(val, a, pos);
+		}
+		return renderPosFloat(val, a, pos);
+	}
+	
+
+	def renderPosFloat(val: float, a: Array<byte>, pos: int) -> int {
+		var bits = u32.view(val);
+		var exponent: u32 = (bits >> FLOAT_MANTISSA_BITS) & ((1u << FLOAT_EXP_WIDTH) - 1);
+		var mantissa: u32 = bits & ((1u << FLOAT_MANTISSA_BITS) - 1);
+
+		if (DEBUG)
+			System.puts(Strings.format2("Exponent: %d | Mantissa: %d \n", exponent, mantissa));
+
+		if (exponent == u32.view(0b11111111) || (exponent == 0 && mantissa == 0)) {
+			return renderFloatSpecial(exponent, mantissa, a, pos);
+		}
+
+		return renderPosFloatNaive(exponent, mantissa, a, pos);
+	}
+
+
+	def renderFloatSpecial(exponent: u32, mantissa: u32, a: Array<byte>, pos: int) -> int {
+		if (mantissa != 0) {
+			a[pos++] = 'N';
+			a[pos++] = 'a';
+			a[pos] = 'N';
+			return 3;
+		}
+		if (exponent != 0) {
```

Comment:
You could use `Arrays.copyInto` for these sequences.

Also side note: I was planning on refactoring both `Ints` and `Longs` to use `Range<byte>` but until then, it's good that you're consistent with them.